src/swarmpal/toolboxes/tfa/tfa_processor.py

```diff
@@ -15,192 +15,662 @@
+        field from the VFM instrument, etc.
+    sampling_time : str
+        String that specifies the time between two consecutive measurements
+        to be given according to the ISO 8601 standard for time intervals
+        e.g. PT1S for one second resolution, PT0.02S for 0.02 seconds,
+        PT1M for one minute resolution etc
+    remove_chaos : bool
+        Set to True to allow a subsequent call of subtract_chaos() to
+        remove the CHAOS-Core and CHAOS-Static field model from the data.
+        It only applies to magnetic field variables.
+    """
+
+    # default values for the class parameters
+    TIME_LIMS = []
+    COLLECTION = "SW_OPER_MAGB_LR_1B"
+    VARNAME = "B_NEC"
+    SAMPLING_TIME = 1
+    FLAGNAME = "Flags_B"
+    MAX_FLAG_VAL = 30
+    REMOVE_CHAOS = True
+
     COLLECTIONS = [
         *[f"SW_OPER_MAG{x}_LR_1B" for x in "ABC"],
         *[f"SW_OPER_MAG{x}_HR_1B" for x in "ABC"],
+        *[f"SW_OPER_EFI{x}_LP_1B" for x in "ABC"],
+        *[f"SW_OPER_IBI{x}TMS_2F" for x in "ABC"],
+        *[f"SW_OPER_FAC{x}TMS_2F" for x in "ABC"],
+        *[f"SW_EXPT_EFI{x}_TCT02" for x in "ABC"],
     ]
 
     DEFAULTS = {
-        "measurements": ["B_NEC", "F"],
-        "model": "'CHAOS-Core' + 'CHAOS-Static'",
+        "measurements": ["B_NEC", "Flags_B"],
+        "model": "None",  # "'CHAOS-Core' + 'CHAOS-Static'" or "None"
         "auxiliaries": ["QDLat", "QDLon", "MLT"],
-        "sampling_step": None,
+        "sampling_step": "PT1S",
         "pad_times": [dt.timedelta(hours=3), dt.timedelta(hours=3)],
     }
 
+    def flag_cleaning(self, varname=None, flagname=None, max_val=None):
+        """Set flagged values to NaN"""
+        # if no inputs are given, default to the class variables
+        if varname is None:
+            varname = self.VARNAME
+        if flagname is None:
+            flagname = self.FLAGNAME
+        if max_val is None:
+            max_val = self.MAX_FLAG_VAL
+
+        inds_to_remove = self.xarray[flagname] > max_val
+        if ("NEC" in self.VARNAME) or ("XYZ" in self.VARNAME):
+            # 3-D case
+            self.xarray[varname][inds_to_remove, :] = np.NaN
+        else:
+            # 1-D case
+            self.xarray[varname][inds_to_remove] = np.NaN
+
+    def calculate_magnitude(self, input_varname=None, output_varname=None):
+        """If requested, calculate the magnitude of a vector variable"""
+        # only run if vector field was given as input
+        if ("NEC" in self.VARNAME) or ("XYZ" in self.VARNAME):
+            # if no inputs are given, default to the class variables
+            if input_varname is None:
+                input_varname = self.VARNAME
+            if output_varname is None:
+                output_varname = "B"
+
+            X = np.sqrt(np.sum(self.xarray[input_varname] ** 2, 1))
+            self.append_array(output_varname, X.data, dims=("Timestamp"))
+        else:
+            logging.warn(
```

Comment:
Looks like these should probably be Exceptions - i.e. program execution should be halted